### DISCUSSION TITLE
TuyaReceive not work some payload

### DISCUSSION THREAD

**Discussion by pptp-plg** [arendst/Tasmota#20086] [Support request] [answered]
### PROBLEM DESCRIPTION

TuyaReceive not work  message 
0x55 0xAA 0x03 0x07 0x00 0x08 0x02 0x02 0x00 0x04 0x00 0x00 0x55 0xDD 0x4B

I think this overflow or checksum calculate incorrect.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): Tuya mcu
- [x] Tasmota binary firmware version number used: 13.2.0
  - [x ] Pre-compiled
  - [ ] Self-compiled
- [ x] Flashing tools used: tas
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by Jason2866** [comment]
Please add more Infos. Which Device? What have you tried?

**Reply by pptp-plg** [reply]
micro solar inverter device.
and I try to send by manual use usb serial it not work. but other payload work fine

0x55 0xAA 0x03 0x07 0x00 0x08 0x66 0x02 0x00 0x04 0x00 0x00 0x00 0x00 0x7D

**Reply by pptp-plg** [reply]
I try to decrease value data on payload proplem 
0x55AA0307000802020004000055DD4B
to
0x55AA03070008020200040000454DAB
it work.
Last data can is 21759 (0x54FF).

**Comment by pptp-plg** [comment]
I found bug if payload have 0x55 in some byte tuyamcu work incorrect

**Reply by barbudor** [reply]
Ok, You are rigth
I see that in the code : Any 0x55 reset the state machine

Let me look at that

Are you using a standard Tasmota build (if yes, which one) or are you able to compile your build yourself ?

**Reply by pptp-plg** [reply]
standard Tasmota 13.2.0

**Reply by barbudor** [reply]
Please test this new binary and report your results
_removed outdated test binary_

Change are [here](https://github.com/barbudor/Tasmota/blob/fix-tyua-state-machine/tasmota/tasmota_xdrv_driver/xdrv_16_tuyamcu_v1.ino) and will be pushed to Tasmota depending on tests results (yours and more)

**Reply by pptp-plg** [reply]
It not work can be test by payload
0x55AA0307000802020004000055DD4B
and
0x55AA03070008020200040000454DAB

**Reply by barbudor** [reply, marked as answer]
Please test this version
[test tuya.zip](https://github.com/arendst/Tasmota/files/13467135/test.tuya.zip)

Looks good to me:
with `55 AA 03 55 00 0A 55 AA 55 AA 55 AA 55 AA 55 AA 5C`:
```
15:44:43.239 {"TuyaReceived":{"Data":"55AA0355000A55AA55AA55AA55AA55AA5C","Cmnd":85,"CmndData":"55AA55AA55AA55AA55AA"}}
15:44:43.241 TYA: RX unknown command
```
with `55 AA 03 55 00 0C 55 AA 55 AA 55 00 AA 55 00 55 AA 00 B4`:
```
15:45:39.852 {"TuyaReceived":{"Data":"55AA0355000C55AA55AA5500AA550055AA00B4","Cmnd":85,"CmndData":"55AA55AA5500AA550055AA00"}}
15:45:39.854 TYA: RX unknown command
```
Commands are meaningless but they are received properly

**Reply by barbudor** [reply]
And with your own examples:
```
16:05:37.066 {"TuyaReceived":{"Data":"55AA0307000802020004000055DD4B","Cmnd":7,"CmndData":"02020004000055DD","DpType2Id2":21981,"2":{"DpId":2,"DpIdType":2,"DpIdData":"000055DD"}}}
16:05:37.069 TYA: fnId=0 is set for dpId=2
16:05:37.071 TYA: RX value 21981 from dpId 2 
16:05:38.908 TYA: Send "55aa00000000ff"
16:05:44.643 {"TuyaReceived":{"Data":"55AA03070008020200040000454DAB","Cmnd":7,"CmndData":"020200040000454D","DpType2Id2":17741,"2":{"DpId":2,"DpIdType":2,"DpIdData":"0000454D"}}}
16:05:44.648 TYA: fnId=0 is set for dpId=2
16:05:44.650 TYA: RX value 17741 from dpId 2 
```

**Reply by barbudor** [reply]
Does it work ?
Without proper confirmation from your side, the fix will NOT be added to tasmota

**Reply by pptp-plg** [reply]
It work.
